eth/protocols/sstorage/peer.go

```
@@ -63,7 +63,10 @@ func (p *Peer) Shards() map[common.Address][]uint64 {
 
 // SetShards this should only be set when doing handshake.
 func (p *Peer) SetShards(shards map[common.Address][]uint64) {
-	p.shards = shards
+	// shards can only be set once.
+	if p.shards != nil {
```

I don't like hiding the issue - it should return an error and the caller should log an error for this.